static/app/views/organizationContext.tsx

```diff
@@ -1,5 +1,147 @@
-import {createContext} from 'react';
+import {Component, createContext, useCallback, useEffect, useRef} from 'react';
+import * as Sentry from '@sentry/react';
 
+import {fetchOrganizationDetails} from 'sentry/actionCreators/organization';
+import {openSudo} from 'sentry/actionCreators/sudoModal';
+import SentryTypes from 'sentry/sentryTypes';
+import ConfigStore from 'sentry/stores/configStore';
+import OrganizationsStore from 'sentry/stores/organizationsStore';
+import OrganizationStore from 'sentry/stores/organizationStore';
+import {useLegacyStore} from 'sentry/stores/useLegacyStore';
 import {Organization} from 'sentry/types';
+import {metric} from 'sentry/utils/analytics';
+import getRouteStringFromRoutes from 'sentry/utils/getRouteStringFromRoutes';
+import useApi from 'sentry/utils/useApi';
+import {useParams} from 'sentry/utils/useParams';
+import {useRoutes} from 'sentry/utils/useRoutes';
 
 export const OrganizationContext = createContext<Organization | null>(null);
+
+interface Props {
+  children: React.ReactNode;
+}
+
+/**
+ * There are still a number of places where we consume the lgacy organization
+ * context. So for now we still need a component that provides this.
+ */
+class LegacyOrganizationContextProvider extends Component<{value: Organization | null}> {
+  static childContextTypes = {
+    organization: SentryTypes.Organization,
+  };
+
+  getChildContext() {
+    return {organization: this.props.value};
+  }
+
+  render() {
+    return this.props.children;
+  }
+}
+/**
+ * Context provider responsible for loading the organization into the
+ * OrganizationStore if it is not already present.
+ */
+export function OrganizationContextProvider({children}: Props) {
+  const api = useApi();
+  const configStore = useLegacyStore(ConfigStore);
+
+  const {organizations} = useLegacyStore(OrganizationsStore);
+  const {organization, error} = useLegacyStore(OrganizationStore);
+
+  const hasMadeFirstFetch = useRef(false);
+
+  const lastOrganization: string | null =
+    configStore.lastOrganization ?? organizations[0]?.slug ?? null;
+
+  const routes = useRoutes();
+  const params = useParams<{orgId?: string}>();
```

Comment:
Dont we always expect orgId to be present here? If we need undefined I would type the value as 
```suggestion
  const params = useParams<{orgId: string|undefined}>();
```

Reply:
No. There are cases (such as account settings) where it's not possible to have an org